src/maxwell-web-view.c

```diff
@@ -0,0 +1,903 @@
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ *
+ * Author: Juan Pablo Ugarte <ugarte@endlessm.com>
+ *
+ */
+
+#include "maxwell.h"
+#include "js-utils.h"
+
+struct _MaxwellWebView
+{
+  GtkOffscreenWindow parent;
+};
+
+typedef struct
+{
+  GtkWidget     *child;
+  GdkWindow     *offscreen; /* child offscreen window */
+  gchar         *id;        /* canvas-id child property */
+  GtkAllocation  alloc;     /* canvas allocation in viewport coordinates */
+} ChildData;
+
+typedef struct
+{
+  GList      *children;  /* List of ChildData */
+  GList      *pixbufs;   /* List of temp pixbufs to handle maxwell:// requests */
+  gboolean    script_loaded;
+} MaxwellWebViewPrivate;
+
+enum
+{
+  CHILD_PROP_0,
+  CHILD_PROP_CANVAS_ID,
+
+  N_CHILD_PROPERTIES
+};
+
+static GParamSpec *child_properties[N_CHILD_PROPERTIES];
+
+G_DEFINE_TYPE_WITH_PRIVATE (MaxwellWebView, maxwell_web_view, WEBKIT_TYPE_WEB_VIEW)
+
+#define RESOURCES_PATH "/com/endlessm/maxwell"
+#define MAXWELL_WEB_VIEW_PRIVATE(d) ((MaxwellWebViewPrivate *) maxwell_web_view_get_instance_private((MaxwellWebView*)d))
+
+static ChildData *
+maxwell_web_view_child_new (GtkWidget *child)
+{
+  ChildData *data = g_slice_new0 (ChildData);
+
+  data->child = g_object_ref_sink (child);
+
+  return data;
+}
+
+static void
+maxwell_web_view_child_free (ChildData *data)
+{
+  if (data == NULL)
+    return;
+
+  if (data->offscreen && data->child)
+    gtk_widget_unregister_window (gtk_widget_get_parent (data->child),
+                                  data->offscreen);
+
+  if (data->offscreen)
+    gdk_window_destroy (data->offscreen);
+
+  g_clear_object (&data->child);
+
+  g_free (data->id);
+
+  g_slice_free (ChildData, data);
+}
+
+#define MWV_DEFINE_CHILD_GETTER(prop, type, cmpstmt) \
+ChildData * \
+get_child_data_by_##prop (MaxwellWebViewPrivate *priv, type prop) \
+{ \
+  GList *l;\
+  for (l = priv->children; l; l = g_list_next (l)) \
+    { \
+      ChildData *data = l->data; \
+      if (cmpstmt) \
+        return data; \
+    } \
+  return NULL; \
+}
+
+MWV_DEFINE_CHILD_GETTER (id, const gchar *, !g_strcmp0 (data->id, id))
+MWV_DEFINE_CHILD_GETTER (child, GtkWidget *, data->child == child)
+MWV_DEFINE_CHILD_GETTER (offscreen, GdkWindow *, data->offscreen == offscreen)
+
+static void
+maxwell_web_view_init (MaxwellWebView *self)
+{
+  MaxwellWebViewPrivate *priv = MAXWELL_WEB_VIEW_PRIVATE (self);
+
+  priv->children = NULL;
+  priv->pixbufs  = NULL;
```

Comment:
These should not be needed, the priv structured is initialized with zeroes anyway.